server/src/main/java/io/druid/server/coordinator/CostBalancerStrategy.java

```diff
@@ -330,6 +330,10 @@ protected double computeCost(
     List<ListenableFuture<Pair<Double, ServerHolder>>> futures = Lists.newArrayList();
 
     for (final ServerHolder server : serverHolders) {
+      if (server.getAvailableSize() < proposalSegment.getSize()) {
```

Comment:
I'm not sure where the loop is. Each round of choosing best server will evaluate it, but serverHolders is bound.

Can you expand more on where you think a loop might occur?